kong/router/fields.lua

```diff
@@ -0,0 +1,236 @@
+local buffer = require("string.buffer")
+local context = require("resty.router.context")
+
+
+local type = type
+local pairs = pairs
+local ipairs = ipairs
+local tb_sort = table.sort
+local tb_concat = table.concat
+local replace_dashes_lower = require("kong.tools.string").replace_dashes_lower
+
+
+local SIMPLE_FIELDS_FUNCS = {
+    -- http.*
+
+    ["http.method"] =
+    function(v, params, cb)
+      return cb("http.method", params.method)
+    end,
+
+    ["http.path"] =
+    function(v, params, cb)
+      return cb("http.path", params.uri)
+    end,
+
+    ["http.host"] =
+    function(v, params, cb)
+      return cb("http.host", params.host)
+    end,
+
+    -- tls.*
+
+    ["tls.sni"] =
+    function(v, params, cb)
+      return cb("tls.sni", params.sni)
+    end,
+
+    -- net.*
+
+    ["net.src.ip"] =
+    function(v, params, cb)
+      return cb("net.src.ip", params.src_ip)
+    end,
+
+    ["net.src.port"] =
+    function(v, params, cb)
+      return cb("net.src.port", params.src_port)
+    end,
+
+    ["net.dst.ip"] =
+    function(v, params, cb)
+      return cb("net.dst.ip", params.dst_ip)
+    end,
+
+    ["net.dst.port"] =
+    function(v, params, cb)
+      return cb("net.dst.port", params.dst_port)
+    end,
+
+    -- below are atc context only
+
+    ["net.protocol"] =
+    function(v, params, cb)
+      return cb("net.protocol", params.scheme)
+    end,
+
+    ["net.port"] =
+    function(v, params, cb)
+      return cb("net.port", params.port)
+    end,
+}
+
+
+local COMPLEX_FIELDS_FUNCS = {
+    ["http.headers."] =
+    function(v, params, cb)
+      local headers = params.headers
+      if not headers then
+        return true
+      end
+
+      for _, name in ipairs(v) do
+        local value = headers[name]
+
+        local res, err = cb("http.headers." .. name, value,
+                            replace_dashes_lower) -- only for cache_key
+        if not res then
+          return nil, err
+        end
+      end -- for ipairs(v)
+
+      return true
+    end,
+
+    ["http.queries."] =
+    function(v, params, cb)
+      local queries = params.queries
+      if not queries then
+        return true
+      end
+
+      for _, name in ipairs(v) do
+        local value = queries[name]
+
+        local res, err = cb("http.queries." .. name, value)
+        if not res then
+          return nil, err
+        end
+      end -- for ipairs(v)
+
+      return true
+    end,
+}
+
+
+-- cache key string
+local str_buf = buffer.new(64)
+
+
+local function get_cache_key(fields, params)
```

Comment:
I might not have made myself clear, the callback should not be on fields. It should be on the `get_cache_key` function instead.

The field callback should only return the field value, like:

```
local buf = buffer.new()
get_params(function(field, v)
  str_buf:put(v or ""):put("|")
end)

local ctx = context.new()
get_params(function(field, v)
  return context:add_value(field, v)
end)
```

Not the other way around as it makes the abstraction much less meaningful.

Reply:
done.